### ISSUE TITLE
npu_export: the gemma3-270m preset's mask concat lands just over a size threshold where the HTP applies it to the wrong rows

### ISSUE THREAD

**Issue by john-rocky** [google-ai-edge/litert-torch#1184]
I've been taking `npu_export` end to end — Stages 1–4 from the bundled gemma3-270m preset, then the bundle on a Galaxy S26 (SM8850 / hexagon v81, QAIRT 2.47). The pipeline itself gets there: all four subgraphs dispatch to the NPU and the model prefills 128 tokens at ~17k tok/s. But the attention mask is applied to the wrong rows once it runs on the HTP. Three of the four query heads end up non-causal during prefill. On a 45-prompt set the bundle answers 12/45, where the same statically-quantized graph run by the TFLite CPU interpreter answers 25/45 and fp32 answers 28/45. (Every score here is from a Stage 2 run with 10 calibration prompts × 32 decode steps, so the arms are comparable; a 50 × 128 budget shifts them by about 2.)

**Changing `cache_length` from 1024 to 896 in `configs/gemma3_270m_qualcomm_sm8850.json` fixes it completely — 29/45 against fp32's 28/45, with no measurable change in decode speed.** I'd be glad to send that as a PR. Equally happy to take the reproducer below to Qualcomm instead if you'd rather it be fixed underneath — whichever is more useful to you.

## What the graph does, and where it diverges

`prefill_128` builds the mask operand by concatenating `prefill_128_mask_local` `[1,1,128,1152]` four times along axis 2 — one copy per query head, since the preceding reshape stacks the heads along rows — and adds it to the attention scores.

I pulled the QNN context out of the tflite Stage 4 produced — 282,308,608 B, the size the runtime reports loading — and ran it standalone under `qnn-net-run`, fed **the exact int16 tensors the runtime feeds it**, dumped out of `litert_lm_advanced_main` with `LITERTLM_DUMP_DIR`.

To read intermediate tensors I promoted a few of them to graph outputs and recompiled. Two checks that the instrumented build is still the same computation: its kv_slice outputs are byte-identical to the uninstrumented compile (36/36), and byte-identical to what the runtime itself produces on the same inputs (36/36).

Against the same graph in the TFLite interpreter, one prompt (25 tokens):

| probe | op | corr | max abs diff |
|---|---|---|---|
| `q_proj` | FULLY_CONNECTED | 1.00000 | 7 |
| `kv_slice_k_0` | TRANSPOSE | 1.00000 | 8 |
| QK^T local | BATCH_MATMUL | 1.00000 | 14 |
| the score concat | CONCATENATION | 1.00000 | 14 |
| **the ADD that applies the mask** | **ADD** | **0.92784** | **28981** |
| softmax | SOFTMAX | 0.92354 | 29414 |
| `kv_slice_k_17` | (layer 17) | 0.67245 | 14433 |

Everything through the matmuls agrees to single-digit LSBs; the ADD disagrees by the mask's own magnitude (tensor RMS is 28,722), and the rest is that error propagating through 18 layers.

Recovering the mask that was actually added — the two probes subtracted and dequantized — says which rows landed where:

| | effective mask row == input mask row *t* |
|---|---|
| TFLite CPU interpreter | 512/512 |
| the compiled graph on the HTP | **367/512** |

Head 0's mask restarts every 16 rows. Heads 1, 2 and 3 receive the final, fully-open mask row for every query position, so on those heads every token can attend to every valid token, including the ones after it. A second prompt (16 tokens) gives the same per-head pattern.

## A two-op reproducer, no weights

984-byte tflite, `apply_plugin` with the same flags Stage 4 uses:

```
concat_out = CONCATENATION([mask, mask, mask, mask], axis=2)   # 4 x [1,1,128,C] -> [1,1,512,C]
add_out    = ADD(scores, concat_out)                           # scores is [1,1,512,C]
```

`mask` carries a row signature (row *t* is 0 for columns ≤ *t*, −32767 beyond) so a wrong row index shows as a wrong open-column count. `scores` is zero. Both tensors are graph outputs.

`concat_out` comes back byte-identical to the CPU interpreter in **every** case below. `add_out` does not, once the concat's output passes 1 MiB:

| bytes | MiB | shape | `concat_out` | `add_out` |
|---:|---:|---|---|---|
| 983,040 | 0.9375 | 512 × 960 | OK | OK |
| 1,048,576 | 1.0000 | 512 × 1024 | OK | **OK** |
| 1,081,344 | 1.0312 | 512 × 1056 | OK | **WRONG** |
| 1,114,112 | 1.0625 | 512 × 1088 | OK | WRONG |
| 1,179,648 | 1.1250 | 256 × 2304 | OK | WRONG |
| 1,179,648 | 1.1250 | 512 × 1152 | OK | WRONG |
| 884,736 | 0.8438 | 384 × 1152 | OK | OK |
| 884,736 | 0.8438 | 512 × 864 | OK | OK |

It tracks total size, not shape: the two 1.125 MiB shapes both fail and the two 0.844 MiB shapes both pass. In the failing cases the ADD pairs output row *r* with the wrong operand row — block 1 reads +32, block 2 +64, block 3 +96.

Three things it is **not**:

- not the repeated operand — concatenating four *distinct* inputs fails identically;
- not the ADD at that size — a plain ADD of two `[1,1,512,1152]` inputs, no concat in the graph, is correct;
- not the optimization level — `-O1`, `-O2` and `-O3` all reproduce it, and on the full model `-O1` and `-O3` give byte-identical outputs.

Whether this sits in the compiler or in the runtime underneath isn't something I can see from outside.

## The preset sits one step over the line

The mask is `[1, 1, prefill, cache_length + prefill]` concatenated once per head, so `cache_length` decides which side it lands on:

| `cache_length` | mask concat | | HTP mask rows correct | 45-prompt score |
|---|---|---|---|---|
| 1024 (the preset) | 512 × 1152 | 1.125 MiB | 367/512 | 12/45 |
| 960 | 512 × 1088 | 1.062 MiB | not probed | 12/45 |
| **896** | 512 × 1024 | **1.000 MiB** | **512/512** | **29/45** |

All three built identically apart from `cache_length`.

960 is the control: it is also a shorter cache, and it does not help — it sits at the smallest failing size in the table above. 896 is one 64-column step below it, and there the mask lands on all 512 rows and the model comes back. A second 896 bundle built from an independent Stage 2 run (50 prompts × 128 decode steps) scores 27/45, and the matching 1024 bundle scores 14/45 — so this isn't calibration variance. Calibration budget moves this set by about 2 answers; `cache_length` moves it by 17.

Decode speed is unchanged between the 1024 and 896 bundles (interleaved A/B, three blocks of three runs alternating, handset warm). The cost is 128 tokens of context.

## Extent measured

`google/gemma-3-270m-it` only, SM8850 / hexagon v81 only, QAIRT 2.47, `-O1`/`-O2`/`-O3`, int16, `axis=2`, ADD as the consumer, the nine shapes above, 45 prompts, greedy. Not tested: other models, other SoCs, other dtypes, other consumer ops, other concat axes. The decode graph was not probed — everything here is prefill. The other bundled preset, `qwen3_0.6b_qualcomm_sm8750.json`, also has `cache_length` 1024, and Qwen3-0.6B has 16 query heads rather than 4 — which would put its mask concat at `[1,1,2048,1152]`, 4.5 MiB. I have not tested that: no SM8750 handset here, and Stage 4 for that model exceeds the memory I can give the container. Worth the same check by someone who can run it.

Environment: macOS arm64 host for Stages 1–3, Python 3.12.13, litert-torch at `8379afb` (still `main` as I write this) plus #1178, without which Stage 2 doesn't run at all. Stage 4 in a Linux container with `ai-edge-litert-sdk-qualcomm` 2.2.0. Calibration: the 10 bundled sample prompts × 32 decode steps unless stated.

<details>
<summary>Reproducer generator (self-contained; writes the tflite)</summary>

```python
"""Builds a minimal int16 tflite that isolates the suspect construct.

  concat_out = CONCATENATION([mask, mask, mask, mask], axis=2)   # 128 rows -> 512
  add_out    = ADD(scores, concat_out)

Both are graph outputs, so the concat can be checked on its own — which the
full model could not do, because there the concat is an interior tensor and
materializing it might change how it is compiled.

Shapes, dtypes and quantization scales are copied from gemma-3-270m-it's
prefill_128 subgraph (op32 / op33 / op34).

Usage: make_min_repro.py <out.tflite> [n_copies] [--distinct] [--cols N]
                         [--no-concat]

--no-concat replaces the concat with a second full-size input, so the ADD is
exercised at the same shape without a concat feeding it.
"""

import sys

import flatbuffers
from ai_edge_litert import schema_py_generated as schema

OUT = sys.argv[1]
NCOPY = int(sys.argv[2]) if len(sys.argv) > 2 else 4
DISTINCT = "--distinct" in sys.argv
NO_CONCAT = "--no-concat" in sys.argv
_c = [a for a in sys.argv if a.startswith("--cols=")]

ROWS = 128
COLS = int(_c[0].split("=")[1]) if _c else 1152
S_SCORES = 0.0007636651280336082
S_MASK = 0.0030518509447574615
S_ADD = 0.003450772725045681


def q(scale):
  p = schema.QuantizationParametersT()
  p.scale = [scale]
  p.zeroPoint = [0]
  p.quantizedDimension = 0
  return p


def tensor(name, shape, scale, buf):
  t = schema.TensorT()
  t.name = name.encode()
  t.shape = shape
  t.type = schema.TensorType.INT16
  t.buffer = buf
  t.quantization = q(scale)
  return t


m = schema.ModelT()
m.version = 3
m.description = b"minimal repro: concat(mask x N, axis=2) + add"

b0 = schema.BufferT()
b0.data = []
m.buffers = [b0]

sg = schema.SubGraphT()
sg.name = b"main"
tensors = []

# inputs
tensors.append(tensor("scores", [1, 1, ROWS * NCOPY, COLS], S_SCORES, 0))
if DISTINCT:
  for i in range(NCOPY):
    tensors.append(tensor(f"mask{i}", [1, 1, ROWS, COLS], S_MASK, 0))
  mask_idx = list(range(1, 1 + NCOPY))
  n_in = 1 + NCOPY
else:
  tensors.append(tensor("mask", [1, 1, ROWS, COLS], S_MASK, 0))
  mask_idx = [1] * NCOPY          # the same tensor, N times -- the suspect construct
  n_in = 2

tensors.append(tensor("concat_out", [1, 1, ROWS * NCOPY, COLS], S_MASK, 0))
tensors.append(tensor("add_out", [1, 1, ROWS * NCOPY, COLS], S_ADD, 0))
concat_i = len(tensors) - 2
add_i = len(tensors) - 1
sg.tensors = tensors
sg.inputs = list(range(n_in))
sg.outputs = [concat_i, add_i]
if NO_CONCAT:
  # second operand becomes a plain full-size input; no concat in the graph
  tensors[1] = tensor("mask", [1, 1, ROWS * NCOPY, COLS], S_MASK, 0)
  sg.inputs = [0, 1]
  sg.outputs = [add_i]

oc_concat = schema.OperatorCodeT()
oc_concat.builtinCode = schema.BuiltinOperator.CONCATENATION
oc_concat.deprecatedBuiltinCode = schema.BuiltinOperator.CONCATENATION
oc_concat.version = 1
oc_add = schema.OperatorCodeT()
oc_add.builtinCode = schema.BuiltinOperator.ADD
oc_add.deprecatedBuiltinCode = schema.BuiltinOperator.ADD
oc_add.version = 1
m.operatorCodes = [oc_concat, oc_add]

o1 = schema.OperatorT()
o1.opcodeIndex = 0
o1.inputs = mask_idx
o1.outputs = [concat_i]
copts = schema.ConcatenationOptionsT()
copts.axis = 2
copts.fusedActivationFunction = schema.ActivationFunctionType.NONE
o1.builtinOptions = copts
o1.builtinOptionsType = schema.BuiltinOptions.ConcatenationOptions

o2 = schema.OperatorT()
o2.opcodeIndex = 1
o2.inputs = [0, concat_i]
o2.outputs = [add_i]
aopts = schema.AddOptionsT()
aopts.fusedActivationFunction = schema.ActivationFunctionType.NONE
o2.builtinOptions = aopts
o2.builtinOptionsType = schema.BuiltinOptions.AddOptions

if NO_CONCAT:
  o2.inputs = [0, 1]
  sg.operators = [o2]
  m.operatorCodes = [oc_add]
  o2.opcodeIndex = 0
else:
  sg.operators = [o1, o2]
m.subgraphs = [sg]

sd = schema.SignatureDefT()
sd.signatureKey = b"main"
sd.subgraphIndex = 0


def tmap(name, idx):
  t = schema.TensorMapT()
  t.name = name.encode()
  t.tensorIndex = idx
  return t


sd.inputs = [tmap(tensors[i].name.decode(), i) for i in sg.inputs]
sd.outputs = ([tmap("add_out", add_i)] if NO_CONCAT
              else [tmap("concat_out", concat_i), tmap("add_out", add_i)])
m.signatureDefs = [sd]

bld = flatbuffers.Builder(1024)
bld.Finish(m.Pack(bld), b"TFL3")
open(OUT, "wb").write(bytes(bld.Output()))
print(f"wrote {OUT}  copies={NCOPY}  distinct={DISTINCT}  "
      f"rows {ROWS}->{ROWS*NCOPY}")
```

Feed it a mask with the row signature described above and zero scores, compile with
`ApplyPlugin(subgraphs_to_compile=None)(..., soc_manufacturer="Qualcomm", soc_model="SM8850",
qualcomm_optimization_level="O3", qualcomm_log_level="off", qualcomm_enable_weight_sharing="true")`,
pull the context out of the DISPATCH_OP custom options, and run it with
`qnn-net-run --backend libQnnHtp.so --retrieve_context ... --use_native_input_files --use_native_output_files`.
</details>

**Comment by shivvaam0001** [comment]
Hi @john-rocky,

Thank you for reaching out to us. We will investigate this further and get back to you.

**Comment by sirakiin** [comment]
@YunanAZ Could you help TAL? Thanks!

**Comment by aravindmurali22** [comment]
Hi @YunanAZ, following up on this thread, is there any recent updates on this? Thanks!
